pkg/services/notification/resource_control.go

```diff
@@ -134,3 +138,195 @@ func GetEmailServiceConfig(ctx context.Context) *pb.EmailServiceConfig {
+			db = db.Where(column+" LIKE ?", "%"+req.SearchWord.GetValue()+"%")
+		}
+	}
+
+	//Step3：get DisplayColumns
+	var displayColumns = dbutil.GetColumnsFromStruct(&models.Notification{})
+	if req.DisplayColumns != nil && req.DisplayColumns[0] != "" {
+		displayColumns = dbutil.GetDisplayColumns(req.GetDisplayColumns(), displayColumns)
+	}
+
+	//Step4：get OrderByStr
+	var sortKeyStr string = models.NfColStatus
+	var reverseStr string = constants.DESC
+	orderByStr := sortKeyStr + " " + reverseStr
+
+	if req.SortKey != nil {
+		sortKeyStr = req.SortKey.GetValue()
+		if req.Reverse != nil {
+			if req.Reverse.GetValue() {
+				reverseStr = constants.DESC
+			} else {
+				reverseStr = constants.ASC
+			}
+		} else {
+			reverseStr = constants.DESC
+		}
+		orderByStr = sortKeyStr + " " + reverseStr
+	}
+
+	return db, displayColumns, orderByStr
+
+}
+
+func DescribeTasks(req *pb.DescribeTasksRequest) ([]*models.Task, uint64, error) {
+
+	var tasks []*models.Task
+	offset := pbutil.GetOffsetFromRequest(req)
+	limit := pbutil.GetLimitFromRequest(req)
+
+	var count uint64
+	var err error
+	db, displayColumns, orderByStr := buildDB4DescTasks(req)
+
+	tasks = make([]*models.Task, 0)
+	err = db.Select(displayColumns).
+		Offset(offset).
+		Limit(limit).
+		Order(orderByStr).
+		Find(&tasks).
+		Count(&count).
+		Error
+	if err != nil {
+		logger.Errorf(nil, "Failed to describe notifications [%v], error: %+v.", req, err)
+		return nil, 0, err
+	}
+
+	if req.DisplayColumns != nil && req.DisplayColumns[0] == "" {
+		return make([]*models.Task, 0), count, err
+	}
+
+	return tasks, count, err
+
+}
+func buildDB4DescTasks(req *pb.DescribeTasksRequest) (*gorm.DB, []string, string) {
+	//Step1：get Filter
+	db := globalcfg.GetInstance().GetDB()
+
+	nfIds := req.NotificationId
+	taskIds := req.TaskId
+	//taskActions := req.TaskAction
+	errorCodes := req.ErrorCode
+	statuses := req.Status
+
+	if nfIds != nil {
+		db = db.Where(models.TaskColNfId+" in (?)", nfIds)
+	}
+	if taskIds != nil {
+		db = db.Where(models.TaskColTaskId+" in (?)", taskIds)
+	}
+	//if taskActions != nil {
+	//	db = db.Where(models.NfColOwner+" in (?)", owners)
+	//}
+	if statuses != nil {
+		db = db.Where(models.TaskColStatus+" in (?)", statuses)
+	}
+
+	if errorCodes != nil {
+		db = db.Where(models.TaskColErrorCode+" in (?)", errorCodes)
+	}
+	//Step2：get SearchWord
+	if req.SearchWord != nil {
+		for _, column := range models.TaskSearchColumns {
+			db = db.Where(column+" LIKE ?", "%"+req.SearchWord.GetValue()+"%")
+		}
+	}
+
+	//Step3：get DisplayColumns
+	var displayColumns = dbutil.GetColumnsFromStruct(&models.Task{})
+	if req.DisplayColumns != nil && req.DisplayColumns[0] != "" {
+		displayColumns = dbutil.GetDisplayColumns(req.GetDisplayColumns(), displayColumns)
+	}
+
+	//Step4：get OrderByStr
+	var sortKeyStr string = models.NfColStatus
+	var reverseStr string = constants.DESC
+	orderByStr := sortKeyStr + " " + reverseStr
+
+	if req.SortKey != nil {
```

Comment:
Here is a little redundance.

Have a look at: 
https://github.com/kubesphere/im/blob/6ad0bde84b609bad9c79ae5ab23a248646165743/pkg/service/im/resource/group_control.go#L216